documentation/library_guide/parallel_api/pass_data_algorithms.rst

```diff
@@ -81,47 +76,65 @@ the buffer were created for the same queue. For example:
     return 0;
   }
 
-Alternatively, use ``std::vector`` with a USM allocator. For example:
+When using device USM, such as allocated by ``malloc_device``, manually copy data to this memory
+before calling oneDPL algorithms, and copy it back once the algorithms have finished execution.
+
+Use std::vector
+-----------------------------
+
+The following examples demonstrate two ways to use the parallel algorithms with ``std::vector``:
+
+* host allocators
+* USM allocators
+
+You can use iterators to host allocated ``std::vector`` data
+as shown in the following example:
 
 .. code:: cpp
 
   #include <oneapi/dpl/execution>
   #include <oneapi/dpl/algorithm>
-  #include <sycl/sycl.hpp>
+  #include <vector>
   int main(){
-    const int n = 1000;
-    auto policy = oneapi::dpl::execution::dpcpp_default;
-    sycl::usm_allocator<int, sycl::usm::alloc::shared> alloc(policy.queue());
-    std::vector<int, decltype(alloc)> vec(n, alloc);
-
-    std::fill(policy, vec.begin(), vec.end(), 42);
-
+    std::vector<int> vec( 1000 );
+    std::fill(oneapi::dpl::execution::dpcpp_default, vec.begin(), vec.end(), 42);
+    // each element of vec equals to 42
     return 0;
   }
 
-When using device USM, such as allocated by ``malloc_device``, manually copy data to this memory
-before calling oneDPL algorithms, and copy it back once the algorithms have finished execution.
-
-Use Host-Side std::vector
------------------------------
+When using iterators to host allocated data, a temporary SYCL buffer is created, and the data
+is copied to this buffer. After processing on a device is complete, the modified data is copied
```

Comment:
I have a couple of comments here:
1) Unfortunately, that example is not quite correct from very beginning. `std::fill` algo has just output data and doesn't  need to have in the buffer the copied data from the host, moreover, uninitialized data :)
2) Also the wording below  is not  quite correct too (I mean `the data is copied to this buffer`), because we have just the "output data" here, so that algo doesn't  need to have in the buffer the copied data from the host, moreover, uninitialized data :)

Reply:
I really like the idea of adding `dpl::begin(std::vector<...>)`, `dpl::end(std::vector<...>)` overloads, and recommending that over iterators from the standard library.  However, that obviously won't happen within the scope of this PR.  

I'm also in favor of extending the `no_init` tags to our `__get_sycl_range` implementations and aligning with sycl on that (and possibly exposing it to the user as described above).

I'll address the other parts of this comment soon.

Reply:
My thought with the `begin`, `end` overloads would be that we suggest to use them for ALL vectors, not only for USM allocated ones.  Since we have full access to allocator information in the vector type, we can make the correct decision for the user about how to treat their data.  With the iterators only, we aren't guaranteed to have allocator type info.
With this information at our disposal, we can copy to buffers in the case of host allocated data, and pass directly to kernels in the case of USM allocated data.  It allows us to have a single recommendation for use of oneDPL with `std::vector`, without caveats.  
I do understand though that it adds steps for users coming from normal parallel policies, and we would need to weigh that tradeoff.

Also, this is not to say we deprecate using bare iterators, just that we recommend using our `begin` and `end` functions instead for best performance.

--

As for `no_init`, you are likely correct that all the information we need is in the algorithm semantics, and this is not necessary to add to the public API.  There is a related change I was thinking of with regard to our internal implementation where our input data "processing" for the dpcpp backend does not align with the SYCL interface for buffer modes & properties exactly, so our internal interface is less expressive than the SYCL interface.  [Here is an example](https://github.com/oneapi-src/oneDPL/blob/d442b291c86b1d44539f2e472bb96a4067ef6e82/include/oneapi/dpl/pstl/hetero/histogram_impl_hetero.h#L133-L136) where we don't have the expressivity that we want.  I

Reply:
If we take `no_init` out of it for now, my initial thought would be to implement them as follows:
`dpl::begin(vec)`:
check the allocator type and either `vec.data()` or `vec.begin()` depending on USM allocator or not.

`dpl::end(vec)`:
check the allocator type and either `vec.data() + vec.size()` or `vec.end()` depending on USM allocator or not.

The rest can be left to our existing implementation.

Reply:
> My thought with the begin, end overloads would be that we suggest to use them for ALL vectors, not only for USM allocated ones. ... Also, this is not to say we deprecate using bare iterators, just that we recommend using our begin and end functions instead for best performance.

I see, thanks for explaining. That makes more sense but needs to be further elaborated. Let's discuss separately from this patch.

Reply:
Agreed, also @MikeDvorskiy will have to weigh in if his intentions are different, I hijacked his comment.

Reply:
> I'll address the other parts of this comment soon.

The other part, I think, is that the example that uses only `std::fill` is not sufficient to illustrate the aspects of host-side data handling explained in the text. So, I guess, a different example is desirable.

Reply:
I think we can replace `fill` with `generate`, and then add `sort`.

Reply:
This works for me.  However, for the example with host-side iterators, I would prefer to call generate with a `par_unseq` policy.  I don't want to encourage the use of host iterators as input to multiple APIs successively.  This is a bad pattern of usage which results in lots of data transfers to and from the device.  
We can have a comment to explain why, and it may be useful to users to show that with USM allocated data they can get performance gains by not thrashing data transfers.  I'll try to put this together.

Reply:
I've made the change to use `generate` and `sort`.  I've just generated the data sequentially on the host, because to do it efficiently on the device (or even with `par[_unseq]` would take more complicated code than should live in an example which is meant to show how to pass data, and I don't think it makes sense to write example code device code we wouldn't recommend people use.